_posts/2017-08-13-gsoc-karl-1.md

```diff
@@ -0,0 +1,183 @@
+> I’d expect the code to be optimizable to use .apply( thisArg, arguments ) throughout.
+> However, in test_deopt the outer ...args gets copied just to be passed into the inner fn.
+> I can verify that the problem disappears if I rename either the ...args of test_deopt or of the fn arrow function.
+
+  [#5656]: https://github.com/babel/babel/issues/5656
+
+### What’s going on here?
+Now what was happening was that this code would generate the following:
+
+```js
+var log = function log() {
+  var _console;
+
+  return (_console = console).log.apply(_console, arguments);
+};
+
+function test_opt() {
+  log.apply(undefined, arguments);
+}
+
+function test_deopt() {
+  for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
+    args[_key] = arguments[_key];
+  }
+
+  var fn = function fn() {
+    return log.apply(undefined, arguments);
+  };
+  fn.apply(undefined, args);
+}
+```
+
+See that for section there? Usually this is needed as the arguments object isn’t a real array — for example, if you tried to run `arguments.slice()`, it would fail miserably.
+However, in this case it’s only being passed to `Function.prototype.apply`. Surprisingly enough, Babel already bothers to optimise this specific case, like in the `test_opt` example above.
+
+### Trying to fix it
+So what did I do? Adding the problem file as a new test case, I tried to see if I could get the output to reflect what I wanted.
+
+<img class="img-responsive" alt="Test failure of modified code" src="/images/blog/2017-08-13-gsoc-karl-1/babel-test.png">
+
+“Why’s the test failing? Surely if I change it a little it will solve itself.”
+
+Despite spamming `make test-only` and modifying the transforms of referenced identifiers within the code, any change just resulted in a different bunch of tests failing instead.
+
+### The Chromium debugger is “fun”
+Miserable, annoyed and confused, I bothered to fire up the Node.js inspector to step through what was going on.
+
+<img class="img-responsive" alt="Using the Chromium debugger" src="/images/blog/2017-08-13-gsoc-karl-1/debugger-cropped.png">
+
+After returning to my computer from a drink break, I’m gladly greeted to my hard disk light thrashing around and a practically hung computer.
+
+<img class="img-responsive" alt="Chromium process using more than 3GB of memory" src="/images/blog/2017-08-13-gsoc-karl-1/chromium-task-manager.png">
+
+Holding my computer together with judicious applications of <kbd>Alt</kbd> + <kbd>SysRq</kbd> + <kbd>F</kbd>, I managed to work through the flow of things¹ and figure out how exactly the code worked.
+
+Even through all that, I still couldn’t see anything why it was deciding to remove this “necessary” (so I thought) code that was being removed with my original fix.
+
+### The actual problem?
+See the error shown above? That entire code in green wasn’t meant to be there, even though it was “expected”.
+
+Basically: the test was broken. _Great._ :/
```

Comment:
If you want to go into more detail (don't have to), you can explain what snapshot tests are, and how we write tests in babel (or just link to https://github.com/babel/babel/blob/7.0/CONTRIBUTING.md#writing-tests) in the beginning? Just that they are input/output or link to another link about the name since could be something people don't know about.

Pros: easy to write (it's automated), Cons: hard to verify, tests are easily wrong
Assertion tests: Pros: easily verified, tests are more accurate, Cons: "harder" to write since manual, lots of setup/checks, slower